File: chatbot/backend/src/llm/groq.ts
import { LLMProvider } from './interface'
import { ChatMessage, LLMConfig } from '../config'
import { parseOpenAIStream } from './openai'

/**
 * Groq Cloud provider — blazing fast inference via custom LPUs.
 *
 * Groq is fully OpenAI-compatible, so this adapter reuses the OpenAI SSE parser.
 *
 * Recommended models (fastest first):
 * - llama-3.1-8b-instant  — Fastest, 8K context, great for RAG chatbots
 * - llama3-70b-8192        — Smarter but still very fast
 * - mixtral-8x7b-32768     — 32K context window
 * - gemma2-9b-it           — Lightweight, fast
 *
 * Free tier: ~30 req/min, generous token limits. Enough for portfolio traffic.
 * Get API key: https://console.groq.com/keys
 */
export class GroqProvider implements LLMProvider {
  readonly name = 'Groq Cloud'

  isConfigured(config: LLMConfig): boolean {
    return !!config.apiKey
  }

  async chat(
    messages: ChatMessage[],
    config: LLMConfig
  ): Promise<ReadableStream<string>> {
    if (!this.isConfigured(config)) {
      throw new Error(
        'Groq API key is not configured. Set GROQ_API_KEY via `npx wrangler secret put GROQ_API_KEY`'
      )
    }

    const body = {
      model: config.model,
      messages: messages.map((m) => ({
        role: m.role,
        content: m.content,
      })),
      max_tokens: config.maxTokens,
      temperature: config.temperature,
      stream: true,
    }

    const response = await fetch(
      'https://api.groq.com/openai/v1/chat/completions',
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${config.apiKey}`,
        },
        body: JSON.stringify(body),
      }
    )

    if (!response.ok) {
      const error = await response.text()
      throw new Error(`Groq API error (${response.status}): ${error}`)
    }

    if (!response.body) {
      throw new Error('Groq response body is null')
    }

    // Groq uses the exact same SSE format as OpenAI
    return parseOpenAIStream(response.body as ReadableStream<Uint8Array>)
  }
}

File: chatbot/backend-python/src/llm/groq.py
"""
Groq Cloud provider — blazing fast inference via custom LPUs.

Groq is fully OpenAI-compatible, so we use the `groq` Python client library.

Recommended models:
- llama-3.1-8b-instant  — Fastest, 8K context, great for RAG chatbots
- llama3-70b-8192        — Smarter but still very fast
- mixtral-8x7b-32768     — 32K context window
- gemma2-9b-it           — Lightweight, fast

Free tier: ~30 req/min, generous token limits. Enough for portfolio traffic.
Get API key: https://console.groq.com/keys
"""

from typing import AsyncGenerator, List

import groq
from groq.types.chat import ChatCompletionChunk

from .interface import LLMProvider, ChatMessage
from ..config import LLMConfig


class GroqProvider(LLMProvider):
    def __init__(self, config: LLMConfig):
        self._config = config
        self._client = groq.AsyncGroq(api_key=config.api_key)

    @property
    def name(self) -> str:
        return "Groq Cloud"

    def is_configured(self) -> bool:
        return bool(self._config.api_key)

    async def chat(
        self,
        messages: List[ChatMessage],
    ) -> AsyncGenerator[str, None]:
        stream = await self._client.chat.completions.create(
            model=self._config.model,
            messages=[{"role": m.role, "content": m.content} for m in messages],
            max_tokens=self._config.max_tokens,
            temperature=self._config.temperature,
            stream=True,
        )

        async for chunk in stream:
            if isinstance(chunk, ChatCompletionChunk):
                delta = chunk.choices[0].delta.content
                if delta:
                    yield delta
